### PR TITLE
Converting selected text to MD link when pasting a URL

### PR DESCRIPTION
**Pull request by Sinharitik589**
Fixes https://github.com/vector-im/element-web/issues/21634

Signed-off-by: Ritik Kumar [sinharitik18112835@gmail.com](mailto:sinharitik18112835@gmail.com)

---
Here's what your changelog entry will look like:

## ✨ Features
 * Converting selected text to MD link when pasting a URL ([\#8242](https://github.com/matrix-org/matrix-react-sdk/pull/8242)). Fixes vector-im/element-web#21634. Contributed by @Sinharitik589.

https://user-images.githubusercontent.com/67551927/161950913-6951c31d-e495-4812-b55c-977f5c9cc640.mp4

Preview: https://pr8242--matrix-react-sdk.netlify.app
⚠️ Do you trust the author of this PR? Maybe this build will steal your keys or give you malware. Exercise caution. Use test accounts.

### CODE REVIEW

File: src/editor/operations.ts [hunk 1]
```diff
@@ -225,7 +225,10 @@ export function formatRangeAsLink(range: Range) {
         replaceRangeAndAutoAdjustCaret(range, newParts, true, prefixLength, suffixLength);
     } else {
         // We set offset to -1 here so that the caret lands between the brackets
-        replaceRangeAndMoveCaret(range, [partCreator.plain("[" + range.text + "]" + "()")], -1);
+        replaceRangeAndMoveCaret(range,
+            text?[partCreator.plain("[" + range.text + "]" + "(" + text + ")")]:
+                [partCreator.plain("[" + range.text + "]" + "()")],
+            -1);
```
Comment (t3chguy) [inline src/editor/operations.ts:231]:
```suggestion
        replaceRangeAndMoveCaret(range, [partCreator.plain("[" + range.text + "]" + "(" + (text ?? "") + ")")],-1);
```

**Review by t3chguy — COMMENTED** [review]
Looks great, could we get a test written for this please?

**Comment by Sinharitik589** [comment]
> Looks great, could we get a test written for this please?
Sure..

**Comment by t3chguy** [comment]
Whatever manner you can accomplish executing your code path in, ideally without making internal methods public

**Comment by t3chguy** [comment]
Hey @Sinharitik589 are you still around? This is an amazing change that I've been wishing was merged for a little while now, if you're struggling with the tests we can land it without and someone else can add tests later!

**Comment by Sinharitik589** [comment]
> Hey @Sinharitik589 are you still around? This is an amazing change that I've been wishing was merged for a little while now, if you're struggling with the tests we can land it without and someone else can add tests 
Just caught up with college examinations.. Will be free in a week or so..But we can land these changes and tests can be added later on

**Review by t3chguy — APPROVED** [review]

**Comment by t3chguy** [comment]
Great, will land it for now - really appreciate this change and good luck for your examinations!

**Comment by Johennes** [comment]
@t3chguy I'm noticing on latest nightly that, without having selected anything, pasting `https://foo.bar` puts `[](https://foo.bar)` into the editor. Is that intended? I think this might be an issue because now you have to manually remove the markup if you want to send a raw URL.

**Comment by t3chguy** [comment]
@Johennes https://github.com/vector-im/element-web/issues/22117

**Pull request merged**